File: src/pipeline/claudePlanner.ts
import type { Planner } from "./types";
import type { PlannedObject, Room, ScenePlan, Vec3 } from "../scene/schema";
import { runClaude } from "../llm/claudeCli";

// Real planner backed by the headless local `claude` CLI. Opus 4.8 is asked to emit the
// plan as a single JSON object, which we validate into a typed ScenePlan. Anything
// off-shape throws — no silent fallback.
const MIN_OBJECTS = 3;
const MAX_OBJECTS = 6;

const SYSTEM_PROMPT = [
  "You are a 3D interior scene planner.",
  "Given a short scene description, design one plausible room and the objects inside it,",
  "then return them as a single JSON object matching the schema below.",
  "",
  "Rules:",
  `- Include between ${MIN_OBJECTS} and ${MAX_OBJECTS} distinct objects — the main furniture/props the description implies.`,
  "- Each meshyPrompt describes ONE isolated object for text-to-3D generation: give its",
  "  silhouette, material, and archetype. No brand names or trademarked/IP characters, no",
  "  background or setting, no other objects, no people — just the single object itself.",
  "- approxSize is the object's bounding box [x, y, z] in meters; use realistic dimensions.",
  "- The room origin is its center, on the floor. Floor is y = 0 and Y points up.",
  "- position is each object's CENTER in world meters: keep it inside the room",
  "  (x within ±width/2, z within ±depth/2) and rest it on the floor (y ≈ approxSize[1] / 2,",
  "  or the appropriate mounting height). Do not overlap objects.",
  "- rotationYDeg is the yaw in degrees; 0 faces +Z. Orient objects sensibly.",
].join("\n");

const VEC3_SCHEMA = { type: "array", items: { type: "number" }, minItems: 3, maxItems: 3 } as const;

// The shape the model must return (formerly the emit_scene_plan tool's input_schema).
// The headless CLI exposes no tool-calling, so the schema is delivered in the prompt
// text and the model replies with a matching JSON object.
const SCENE_PLAN_SCHEMA = {
  type: "object",
  additionalProperties: false,
  required: ["room", "objects"],
  properties: {
    room: {
      type: "object",
      additionalProperties: false,
      required: ["width", "depth", "height"],
      properties: {
        width: { type: "number", description: "interior size along X, meters" },
        depth: { type: "number", description: "interior size along Z, meters" },
        height: { type: "number", description: "ceiling height along Y, meters" },
      },
    },
    objects: {
      type: "array",
      minItems: MIN_OBJECTS,
      maxItems: MAX_OBJECTS,
      description: `${MIN_OBJECTS} to ${MAX_OBJECTS} objects placed in the room`,
      items: {
        type: "object",
        additionalProperties: false,
        required: ["id", "label", "meshyPrompt", "approxSize", "position", "rotationYDeg"],
        properties: {
          id: { type: "string", description: "unique kebab-case id, e.g. 'sofa-1'" },
          label: { type: "string", description: "short human label, e.g. 'Sofa'" },
          meshyPrompt: {
            type: "string",
            description:
              "Single isolated object for text-to-3D: silhouette + material + archetype. " +
              "No IP/brand names, no background, no other objects.",
          },
          approxSize: { ...VEC3_SCHEMA, description: "intended bounding box [x, y, z] in meters" },
          position: { ...VEC3_SCHEMA, description: "object center [x, y, z] in world meters (floor at y=0)" },
          rotationYDeg: { type: "number", description: "yaw in degrees; 0 faces +Z" },
        },
      },
    },
  },
};

export const claudePlanner: Planner = {
  async plan(prompt: string): Promise<ScenePlan> {
    const trimmed = prompt.trim();
    if (trimmed.length === 0) {
      throw new Error("claudePlanner.plan: prompt must be a non-empty string");
    }

    const fullPrompt = [
      SYSTEM_PROMPT,
      "",
      `Scene description: ${trimmed}`,
      "",
      "Respond with ONLY the JSON object — no prose, no markdown code fences. It must match this schema:",
      JSON.stringify(SCENE_PLAN_SCHEMA, null, 2),
    ].join("\n");

    const text = await runClaude(fullPrompt);

    const input = asRecord(parseJson(text), "scene plan");
    const roomRaw = asRecord(input.room, "room");
    const room: Room = {
      width: asNumber(roomRaw.width, "room.width"),
      depth: asNumber(roomRaw.depth, "room.depth"),
      height: asNumber(roomRaw.height, "room.height"),
    };

    if (!Array.isArray(input.objects)) {
      throw new Error(`claudePlanner: expected objects[] array, got ${describe(input.objects)}`);
    }
    if (input.objects.length < MIN_OBJECTS || input.objects.length > MAX_OBJECTS) {
      throw new Error(
        `claudePlanner: expected ${MIN_OBJECTS}-${MAX_OBJECTS} objects, got ${input.objects.length}`,
      );
    }

    const seenIds = new Set<string>();
    const objects: PlannedObject[] = input.objects.map((raw, i) => {
      const o = asRecord(raw, `objects[${i}]`);
      const id = asNonEmptyString(o.id, `objects[${i}].id`);
      if (seenIds.has(id)) {
        throw new Error(`claudePlanner: duplicate object id "${id}"`);
      }
      seenIds.add(id);
      return {
        id,
        label: asNonEmptyString(o.label, `objects[${i}].label`),
        meshyPrompt: asNonEmptyString(o.meshyPrompt, `objects[${i}].meshyPrompt`),
        approxSize: asVec3(o.approxSize, `objects[${i}].approxSize`),
        position: asVec3(o.position, `objects[${i}].position`),
        rotationYDeg: asNumber(o.rotationYDeg, `objects[${i}].rotationYDeg`),
      };
    });

    return { prompt, room, objects };
  },
};

// --- response parsing + validation: the model's output is untrusted, so narrow loudly ---

// The CLI returns the model's final text; it may wrap the JSON in a ```json fence.
// Strip a fenced block if present, then parse — failing loudly on anything unparseable.
function parseJson(text: string): unknown {
  let body = text.trim();
  const fence = /```(?:json)?\s*\n?([\s\S]*?)```/.exec(body);
  if (fence) {
    body = fence[1].trim();
  }
  try {
    return JSON.parse(body);
  } catch (cause) {
    throw new Error(`claudePlanner: model output was not valid JSON: ${body.slice(0, 1000)}`, { cause });
  }
}

function asRecord(value: unknown, ctx: string): Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new Error(`claudePlanner: expected an object at ${ctx}, got ${describe(value)}`);
  }
  return value as Record<string, unknown>;
}

function asNumber(value: unknown, ctx: string): number {
  if (typeof value !== "number" || !Number.isFinite(value)) {
    throw new Error(`claudePlanner: expected a finite number at ${ctx}, got ${describe(value)}`);
  }
  return value;
}

function asNonEmptyString(value: unknown, ctx: string): string {
  if (typeof value !== "string" || value.trim().length === 0) {
    throw new Error(`claudePlanner: expected a non-empty string at ${ctx}, got ${describe(value)}`);
  }
  return value;
}

function asVec3(value: unknown, ctx: string): Vec3 {
  if (!Array.isArray(value) || value.length !== 3) {
    throw new Error(`claudePlanner: expected a [x, y, z] array at ${ctx}, got ${describe(value)}`);
  }
  return [
    asNumber(value[0], `${ctx}[0]`),
    asNumber(value[1], `${ctx}[1]`),
    asNumber(value[2], `${ctx}[2]`),
  ];
}

function describe(value: unknown): string {
  if (value === null) return "null";
  if (Array.isArray(value)) return `array(len ${value.length})`;
  return typeof value;
}

File: backend/src/pipeline/claude_planner.py
"""Claude-backed scene planner for Dream3D."""

import json
import math
import re
from typing import Any

from llm.claude_cli import run_claude
from pipeline.prompts import load
from scene.schema import PlannedObject, Room, ScenePlan, Vec3

MIN_OBJECTS = 3
MAX_OBJECTS = 6

SYSTEM_PROMPT = load("planner_system.md").replace("{min_objects}", str(MIN_OBJECTS)).replace("{max_objects}", str(MAX_OBJECTS))

VEC3_SCHEMA = {
    "type": "array",
    "items": {"type": "number"},
    "minItems": 3,
    "maxItems": 3,
}

SCENE_PLAN_SCHEMA = {
    "type": "object",
    "additionalProperties": False,
    "required": ["room", "objects"],
    "properties": {
        "room": {
            "type": "object",
            "additionalProperties": False,
            "required": ["width", "depth", "height"],
            "properties": {
                "width": {
                    "type": "number",
                    "description": "interior size along X, meters",
                },
                "depth": {
                    "type": "number",
                    "description": "interior size along Z, meters",
                },
                "height": {
                    "type": "number",
                    "description": "ceiling height along Y, meters",
                },
            },
        },
        "objects": {
            "type": "array",
            "minItems": MIN_OBJECTS,
            "maxItems": MAX_OBJECTS,
            "description": f"{MIN_OBJECTS} to {MAX_OBJECTS} objects placed in the room",
            "items": {
                "type": "object",
                "additionalProperties": False,
                "required": [
                    "id",
                    "label",
                    "meshyPrompt",
                    "approxSize",
                    "position",
                    "rotationYDeg",
                ],
                "properties": {
                    "id": {
                        "type": "string",
                        "description": "unique kebab-case id, e.g. 'sofa-1'",
                    },
                    "label": {
                        "type": "string",
                        "description": "short human label, e.g. 'Sofa'",
                    },
                    "meshyPrompt": {
                        "type": "string",
                        "description": (
                            "Single isolated object for text-to-3D: silhouette + material + archetype. "
                            "No IP/brand names, no background, no other objects."
                        ),
                    },
                    "approxSize": {
                        **VEC3_SCHEMA,
                        "description": "intended bounding box [x, y, z] in meters",
                    },
                    "position": {
                        **VEC3_SCHEMA,
                        "description": "object center [x, y, z] in world meters (floor at y=0)",
                    },
                    "rotationYDeg": {
                        "type": "number",
                        "description": "yaw in degrees; 0 faces +Z",
                    },
                },
            },
        },
    },
}


async def plan(prompt: str) -> ScenePlan:
    trimmed = prompt.strip()
    if len(trimmed) == 0:
        raise ValueError("claudePlanner.plan: prompt must be a non-empty string")

    full_prompt = "\n".join(
        [
            SYSTEM_PROMPT,
            "",
            f"Scene description: {trimmed}",
            "",
            "Respond with ONLY the JSON object — no prose, no markdown code fences. It must match this schema:",
            json.dumps(SCENE_PLAN_SCHEMA, indent=2),
        ]
    )

    text = run_claude(full_prompt, caller="planner")
    input_ = _as_record(_parse_json(text), "scene plan")
    room_raw = _as_record(input_.get("room"), "room")
    room = Room(
        width=_as_number(room_raw.get("width"), "room.width"),
        depth=_as_number(room_raw.get("depth"), "room.depth"),
        height=_as_number(room_raw.get("height"), "room.height"),
    )

    raw_objects = input_.get("objects")
    if not isinstance(raw_objects, list):
        raise ValueError(
            f"claudePlanner: expected objects[] array, got {_describe(raw_objects)}"
        )
    if len(raw_objects) < MIN_OBJECTS or len(raw_objects) > MAX_OBJECTS:
        raise ValueError(
            f"claudePlanner: expected {MIN_OBJECTS}-{MAX_OBJECTS} objects, got {len(raw_objects)}"
        )

    seen_ids: set[str] = set()
    objects: list[PlannedObject] = []
    for i, raw in enumerate(raw_objects):
        o = _as_record(raw, f"objects[{i}]")
        object_id = _as_non_empty_string(o.get("id"), f"objects[{i}].id")
        if object_id in seen_ids:
            raise ValueError(f'claudePlanner: duplicate object id "{object_id}"')
        seen_ids.add(object_id)
        objects.append(
            PlannedObject(
                id=object_id,
                label=_as_non_empty_string(o.get("label"), f"objects[{i}].label"),
                meshyPrompt=_as_non_empty_string(
                    o.get("meshyPrompt"), f"objects[{i}].meshyPrompt"
                ),
                approxSize=_as_vec3(o.get("approxSize"), f"objects[{i}].approxSize"),
                position=_as_vec3(o.get("position"), f"objects[{i}].position"),
                rotationYDeg=_as_number(
                    o.get("rotationYDeg"), f"objects[{i}].rotationYDeg"
                ),
            )
        )

    return ScenePlan(prompt=prompt, room=room, objects=objects)


def _parse_json(text: str) -> Any:
    body = text.strip()
    fence = re.search(r"```(?:json)?\s*\n?([\s\S]*?)```", body)
    if fence:
        body = fence.group(1).strip()
    try:
        return json.loads(body)
    except json.JSONDecodeError as exc:
        raise ValueError(
            f"claudePlanner: model output was not valid JSON: {body[:1000]}"
        ) from exc


def _as_record(value: Any, ctx: str) -> dict[str, Any]:
    if not isinstance(value, dict):
        raise ValueError(
            f"claudePlanner: expected an object at {ctx}, got {_describe(value)}"
        )
    return value


def _as_number(value: Any, ctx: str) -> float:
    if (
        isinstance(value, bool)
        or not isinstance(value, (int, float))
        or not math.isfinite(value)
    ):
        raise ValueError(
            f"claudePlanner: expected a finite number at {ctx}, got {_describe(value)}"
        )
    return float(value)


def _as_non_empty_string(value: Any, ctx: str) -> str:
    if not isinstance(value, str) or len(value.strip()) == 0:
        raise ValueError(
            f"claudePlanner: expected a non-empty string at {ctx}, got {_describe(value)}"
        )
    return value


def _as_vec3(value: Any, ctx: str) -> Vec3:
    if not isinstance(value, list) or len(value) != 3:
        raise ValueError(
            f"claudePlanner: expected a [x, y, z] array at {ctx}, got {_describe(value)}"
        )
    return (
        _as_number(value[0], f"{ctx}[0]"),
        _as_number(value[1], f"{ctx}[1]"),
        _as_number(value[2], f"{ctx}[2]"),
    )


def _describe(value: Any) -> str:
    if value is None:
        return "null"
    if isinstance(value, list):
        return f"array(len {len(value)})"
    if isinstance(value, bool):
        return "boolean"
    if isinstance(value, str):
        return "string"
    if isinstance(value, (int, float)):
        return "number"
    return type(value).__name__
